# SESIÓN 3: Seguridad y Autenticación

**Objetivo:** Implementar autenticación segura en APIs FastAPI usando OAuth2, JWT y hash de contraseñas con bcrypt.

**Duración:** 180 minutos

**Contenido:**
1. Introducción a OAuth2 y JWT
2. Hash de Contraseñas con bcrypt
3. Generación de Tokens JWT
4. OAuth2PasswordBearer en FastAPI
5. Protección de Rutas

## CONFIGURACIÓN DEL ENTORNO

Ejecuta estas celdas para preparar el entorno de trabajo.

In [ ]:
# Verificación de Python
import sys
print(f"Python: {sys.version}")
assert sys.version_info >= (3, 8), "Se requiere Python 3.8 o superior"

In [1]:
# Instalación de dependencias de seguridad
!pip install fastapi==0.115.0 uvicorn[standard]==0.32.0 passlib[bcrypt]==1.7.4 bcrypt==4.0.1 PyJWT==2.8.0 python-multipart==0.0.6 -q
print("✅ Dependencias instaladas: FastAPI, passlib (bcrypt), PyJWT, python-multipart")
print("\n📝 Nota técnica:")
print("  - PyJWT: Librería moderna para JWT (reemplaza a python-jose que está abandonado)")
print("  - passlib: Estándar para hashing de passwords (bcrypt backend)")

✅ Dependencias instaladas: FastAPI, passlib (bcrypt), PyJWT, python-multipart

📝 Nota técnica:
  - PyJWT: Librería moderna para JWT (reemplaza a python-jose que está abandonado)
  - passlib: Estándar para hashing de passwords (bcrypt backend)


In [2]:
# Imports globales
from fastapi import FastAPI, Depends, HTTPException, status
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from fastapi.testclient import TestClient
from passlib.context import CryptContext
import jwt  # PyJWT
from jwt.exceptions import PyJWTError
from datetime import datetime, timedelta
from typing import Annotated, Optional
from pydantic import BaseModel
import os

print("✅ Imports completados")

✅ Imports completados


### Servidor

Una función para ejecutar el servidor sin tener que reiniciar el kernel

In [3]:
import uvicorn
import threading
import time
import logging

# Variable global para guardar la referencia al servidor activo
# (Esto sobrevive entre ejecuciones de celdas)
if "active_server" not in globals():
    active_server = None


def run_api(app_instance, port=8000):
    global active_server

    logging.getLogger("uvicorn").setLevel(logging.CRITICAL)
    logging.getLogger("uvicorn.error").setLevel(logging.CRITICAL)
    logging.getLogger("uvicorn.access").setLevel(logging.CRITICAL)

    # 1. SI YA HAY UN SERVIDOR, LO APAGAMOS
    if active_server:
        active_server.should_exit = True
        active_server.force_exit = True

        # Esperamos a que libere el puerto (max 3 segundos)
        for _ in range(30):
            if not active_server.started:
                break
            time.sleep(0.1)

    # 2. CONFIGURAMOS EL NUEVO
    config = uvicorn.Config(
        app=app_instance, host="127.0.0.1", port=port, log_level="warning"
    )
    server = uvicorn.Server(config)

    # Guardamos la referencia global
    active_server = server

    # 3. ARRANCAMOS EN UN HILO APARTE
    thread = threading.Thread(target=server.run)
    thread.daemon = True
    thread.start()

    time.sleep(1)
    print(f"🚀 Servidor REINICIADO en http://127.0.0.1:{port}")
    print(f"📄 Documentación: http://127.0.0.1:{port}/docs")


print("✅ Definida corrrectamente")

✅ Definida corrrectamente


---

## 1. INTRODUCCIÓN A OAUTH2 Y JWT

### ¿Qué es OAuth2?

**OAuth2** es un protocolo de autorización (no autenticación) que permite a aplicaciones obtener acceso limitado a cuentas de usuario.

**Flujo típico en una API:**

```
1. Usuario envía credenciales (username/password)
   ↓
2. Servidor valida credenciales
   ↓
3. Servidor genera TOKEN (JWT)
   ↓
4. Cliente guarda token
   ↓
5. Cliente envía token en cada request
   (Header: Authorization: Bearer <token>)
   ↓
6. Servidor valida token y procesa request
```

### ¿Qué es JWT (JSON Web Token)?

JWT es un estándar para crear tokens de acceso que contienen información del usuario.

**Estructura de un JWT:**

```
eyJhbGc.eyJzdWI.SflKxwRJ
   │      │       │
 Header Payload Signature
```

**Partes del JWT:**

1. **Header:** Algoritmo de firma (ej: HS256)
2. **Payload:** Datos del usuario (claims)
   - `sub`: Subject (user_id)
   - `exp`: Expiration time
   - Datos personalizados (email, roles, etc.)
3. **Signature:** Firma criptográfica con SECRET_KEY

### ¿Por qué JWT para APIs?

| Ventaja | Descripción |
|---------|-------------|
| **Stateless** | No requiere almacenar sesiones en servidor |
| **Autónomo** | Contiene toda la información necesaria |
| **Portable** | Funciona entre dominios y servicios |
| **Seguro** | Firmado criptográficamente |
| **Escalable** | No requiere BD de sesiones |

### Ejemplo Visual: Decodificar un JWT

Vamos a ver cómo se ve un JWT decodificado (sin validar la firma).

In [15]:
# Decodificar un JWT de ejemplo (SIN verificar firma)
token_ejemplo = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ1c2VyMTIzIiwiZXhwIjoxNzM5MjI2MDA0LCJlbWFpbCI6InVzZXJAZXhhbXBsZS5jb20ifQ.2xT8pZ9QZ3YxQj7jW8mN5pK4vR2sL1qH6fB9wD0eU3c"

# Decodificar SIN validar (solo para ver contenido)
payload = jwt.decode(token_ejemplo, options={"verify_signature": False}, algorithms=["HS256"])
print("📦 Contenido del JWT (payload):")
print(f"  - sub (user_id): {payload.get('sub')}")
print(f"  - exp (expira): {payload.get('exp')}")
print(f"  - email: {payload.get('email')}")
print("\n⚠️ En producción SIEMPRE verificar la firma")

📦 Contenido del JWT (payload):
  - sub (user_id): user123
  - exp (expira): 1739226004
  - email: user@example.com

⚠️ En producción SIEMPRE verificar la firma


### Micro-reto: Analogía del JWT

Piensa en un JWT como un **pase VIP con holograma de seguridad**:
- Contiene tu información (nombre, nivel de acceso)
- Tiene fecha de caducidad
- Está firmado de forma imposible de falsificar
- No necesitas consultar una base de datos cada vez que lo presentas

In [ ]:
# TODO: ¿Qué pasa si modificamos el payload de un JWT?
# Intenta cambiar el 'sub' del token_ejemplo y vuelve a decodificarlo
# ¿La firma sigue siendo válida?

# Pista: La firma CAMBIA si modificas el payload
# Por eso JWT es seguro contra manipulación

---

## 2. HASH DE CONTRASEÑAS CON BCRYPT

### ⚠️ REGLA DE ORO: NUNCA guardar contraseñas en texto plano

**Problema:**
```python
# ❌ MAL - Contraseña visible
usuarios = {
    "user1": {"password": "mipassword123"}  # Si roban la BD, roban las contraseñas
}
```

**Solución: Hash con bcrypt**
```python
# ✅ BIEN - Hash irreversible
usuarios = {
    "user1": {"password": "$2b$12$KIX..."}
}
```

### ¿Qué es bcrypt?

**bcrypt** es un algoritmo de hash diseñado específicamente para contraseñas:

| Característica | Beneficio |
|----------------|----------|
| **Lento** | Dificulta ataques de fuerza bruta |
| **Salt automático** | Dos contraseñas iguales generan hashes diferentes |
| **Ajustable** | Puedes aumentar la complejidad con el tiempo |
| **Irreversible** | No se puede obtener la contraseña original |

### Conceptos clave

- **Hash:** Transformación unidireccional (password → hash)
- **Salt:** Valor aleatorio que se añade antes de hashear
- **Verificación:** Hashear el intento y comparar con el hash guardado

### Configuración de passlib con bcrypt

Vamos a crear un contexto de hash que usaremos en toda la aplicación.

In [16]:
# Configurar el contexto de hashing con bcrypt
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

print("✅ Contexto de hash configurado con bcrypt")

✅ Contexto de hash configurado con bcrypt


### Ejemplo: Hashear una contraseña

In [23]:
# Contraseña original
password_original = "miSuperPassword123"

# Generar hash
password_hash = pwd_context.hash(password_original)

print("🔒 Hash generado:")
print(f"  Original: {password_original}")
print(f"  Hash: {password_hash}")
print(f"  Longitud: {len(password_hash)} caracteres")
print("\n⚠️ El hash es IRREVERSIBLE")

🔒 Hash generado:
  Original: miSuperPassword123
  Hash: $2b$12$t6d6pmT4hC0pLE46kZo92.jsrgr6RWtsXVfeuuNSI9cDwlfSefdVS
  Longitud: 60 caracteres

⚠️ El hash es IRREVERSIBLE


### Ejemplo: Verificar una contraseña

In [24]:
# Verificar contraseña correcta
es_correcta = pwd_context.verify("miSuperPassword123", password_hash)
print(f"✅ Contraseña correcta: {es_correcta}")

# Verificar contraseña incorrecta
es_incorrecta = pwd_context.verify("otraPassword", password_hash)
print(f"❌ Contraseña incorrecta: {es_incorrecta}")

✅ Contraseña correcta: True
❌ Contraseña incorrecta: False


### Micro-reto: Salt automático

In [25]:
# TODO: Genera 3 hashes de la MISMA contraseña "password123"
# Observa que los hashes son DIFERENTES cada vez
# Esto es gracias al salt automático de bcrypt

password = "password123"
hash1 = pwd_context.hash(password)
hash2 = pwd_context.hash(password)
hash3 = pwd_context.hash(password)

print("Hash 1:", hash1)
print("Hash 2:", hash2)
print("Hash 3:", hash3)

print("\n¿Son iguales?", hash1 == hash2)
print("¿Son iguales?", hash2 == hash3)
print("¿Son iguales?", hash1 == hash3)

print("\nPero TODOS verifican correctamente:", pwd_context.verify(password, hash1))
print("Pero TODOS verifican correctamente:", pwd_context.verify(password, hash2))
print("Pero TODOS verifican correctamente:", pwd_context.verify(password, hash3))

Hash 1: $2b$12$v3ggW3FfvItaHt9vjMd6TulT5rJbR2PM0w0Fuc2UPjS76BnskgZte
Hash 2: $2b$12$dlSsDSir1G/FjphunNipcu.UlMRg3mldhg07k/PrENn5.omdOcTxK
Hash 3: $2b$12$HmpVn4FtFI/I2W7Nv/p6pumSzrgB3vo/43dRkB9jOY88jxTbITgoS

¿Son iguales? False
¿Son iguales? False
¿Son iguales? False

Pero TODOS verifican correctamente: True
Pero TODOS verifican correctamente: True
Pero TODOS verifican correctamente: True


### Funciones reutilizables para toda la app

In [26]:
# Funciones helper que usaremos en la app

def hash_password(password: str) -> str:
    """Hashea una contraseña."""
    return pwd_context.hash(password)

def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verifica si una contraseña coincide con su hash."""
    return pwd_context.verify(plain_password, hashed_password)

# Ejemplo de uso
hash_guardado = hash_password("secreto")
print(f"Hash guardado en BD: {hash_guardado}")
print(f"Contraseña 'secreto' es válida: {verify_password('secreto', hash_guardado)}")
print(f"Contraseña 'incorrecto' es válida: {verify_password('incorrecto', hash_guardado)}")

Hash guardado en BD: $2b$12$.hH73uHuc.sgCl2ztmikdOJQ3fYA0QY.x30Gjb.AoRaPjb8nlRrmy
Contraseña 'secreto' es válida: True
Contraseña 'incorrecto' es válida: False


---

## 3. GENERACIÓN DE TOKENS JWT

### Configuración necesaria

Para generar JWT necesitamos:
1. **SECRET_KEY:** Clave secreta para firmar tokens
2. **ALGORITHM:** Algoritmo de firma (HS256 es el más común)
3. **ACCESS_TOKEN_EXPIRE_MINUTES:** Tiempo de vida del token

### ⚠️ IMPORTANTE: Variables de entorno en producción

**NUNCA** pongas SECRET_KEY en el código en producción.

**Forma CORRECTA:**
```python
import os
SECRET_KEY = os.getenv("SECRET_KEY", "fallback-inseguro-solo-desarrollo")
```

**En el servidor:**
```bash
export SECRET_KEY="tu-clave-super-segura-generada-aleatoriamente"
# O usar un archivo .env con python-dotenv
```

Para este notebook educativo usaremos una clave hardcodeada.

In [27]:
# Configuración de JWT
# ⚠️ SOLO PARA DESARROLLO/EDUCACIÓN
# En producción: SECRET_KEY = os.getenv("SECRET_KEY")
SECRET_KEY = "mi-clave-super-secreta-que-nadie-debe-saber-12345"
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 30

print("✅ Configuración JWT:")
print(f"  - Algoritmo: {ALGORITHM}")
print(f"  - Expiración: {ACCESS_TOKEN_EXPIRE_MINUTES} minutos")
print("\n⚠️ En producción:")
print("  SECRET_KEY = os.getenv('SECRET_KEY')")
print("  Y configurar la variable de entorno en el servidor")

✅ Configuración JWT:
  - Algoritmo: HS256
  - Expiración: 30 minutos

⚠️ En producción:
  SECRET_KEY = os.getenv('SECRET_KEY')
  Y configurar la variable de entorno en el servidor


### Estructura del Payload

El payload es el diccionario de datos que guardamos en el token:

| Campo | Descripción | Ejemplo |
|-------|-------------|----------|
| `sub` | Subject (identificador del usuario) | `"user123"` |
| `exp` | Expiration time (timestamp) | `1739226004` |
| `email` | Email del usuario (custom claim) | `"user@example.com"` |
| `role` | Rol del usuario (custom claim) | `"admin"` |

**Claims estándar:**
- `sub`, `exp`, `iat` (issued at), `iss` (issuer), `aud` (audience)

**Claims personalizados:**
- Cualquier dato que necesites (email, roles, permisos, etc.)

### Función para crear tokens

In [28]:
from datetime import timezone
# type hints 
def crear_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
    """
    Crea un token JWT.
    
    Args:
        data: Diccionario con los claims a incluir (ej: {"sub": "user123"})
        expires_delta: Tiempo de vida del token (opcional)
    
    Returns:
        Token JWT firmado
    """
    to_encode = data.copy()
    
    # Calcular tiempo de expiración
    if expires_delta:
        expire = datetime.now(timezone.utc) + expires_delta # type: ignore
    else:
        expire = datetime.now(timezone.utc) + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES) # type: ignore
    
    # Añadir expiración al payload
    to_encode.update({"exp": expire})
    
    # Crear y firmar el token
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    
    return encoded_jwt

print("✅ Función crear_access_token definida")

✅ Función crear_access_token definida


### Ejemplo: Crear un token

In [29]:
# Datos del usuario
usuario_data = {
    "sub": "user123",
    "email": "user@example.com",
    "role": "admin"
}

# Crear token
token = crear_access_token(usuario_data)

print("🎫 Token generado:")
print(token)
print(f"\nLongitud: {len(token)} caracteres")

# Decodificar para ver contenido (sin verificar firma)
payload_decodificado = jwt.decode(token, options={"verify_signature": False})
print("\n📦 Contenido del token:")
for key, value in payload_decodificado.items():
    print(f"  - {key}: {value}")

🎫 Token generado:
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ1c2VyMTIzIiwiZW1haWwiOiJ1c2VyQGV4YW1wbGUuY29tIiwicm9sZSI6ImFkbWluIiwiZXhwIjoxNzcxNTE0MjYzfQ.KC9rI-icaVTzaNcWeybWwJcPhY2E42FDjcpNfEFPhYI

Longitud: 183 caracteres

📦 Contenido del token:
  - sub: user123
  - email: user@example.com
  - role: admin
  - exp: 1771514263


### Función para decodificar y validar tokens

In [30]:
def decodificar_token(token: str) -> dict:
    """
    Decodifica y valida un token JWT.
    
    Args:
        token: Token JWT a validar
    
    Returns:
        Payload del token si es válido
    
    Raises:
        HTTPException: Si el token es inválido o expirado
    """
    try:
        # Decodificar Y verificar firma
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        return payload
    except PyJWTError as e:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Token inválido o expirado",
            headers={"WWW-Authenticate": "Bearer"},
        )

print("✅ Función decodificar_token definida")

✅ Función decodificar_token definida


### Ejemplo: Validar un token

In [31]:
# Crear token
token_valido = crear_access_token({"sub": "testuser"})

# Validar token correcto
try:
    payload = decodificar_token(token_valido)
    print("✅ Token válido:")
    print(f"  - Usuario: {payload.get('sub')}")
except HTTPException as e:
    print(f"❌ Error: {e.detail}")

# Intentar validar token inválido
token_invalido = "token.falso.aqui"
try:
    decodificar_token(token_invalido)
except HTTPException as e:
    print(f"\n❌ Token inválido rechazado: {e.detail}")

✅ Token válido:
  - Usuario: testuser

❌ Token inválido rechazado: Token inválido o expirado


### Micro-reto: Token expirado

In [32]:
# TODO: Crear un token que expire en 1 segundo
# Esperar 2 segundos e intentar validarlo
# Observa que jwt.decode lanza PyJWTError por expiración

import time

# Crear token que expira en 1 segundo
token_corto = crear_access_token(
    {"sub": "usuario_temporal"},
    expires_delta=timedelta(seconds=1)
)

print("Token creado, esperando 2 segundos...")
time.sleep(2)

try:
    decodificar_token(token_corto)
    print("✅ Token aún válido")
except HTTPException as e:
    print(f"❌ Token expirado: {e.detail}")

Token creado, esperando 2 segundos...
❌ Token expirado: Token inválido o expirado


---

## 4. OAUTH2PASSWORDBEARER EN FASTAPI

### ¿Qué es OAuth2PasswordBearer?

`OAuth2PasswordBearer` es una dependencia de FastAPI que:
1. Lee el header `Authorization: Bearer <token>`
2. Extrae el token
3. Lo pasa a tu dependencia para validarlo

### Flujo completo

```
Cliente                    Servidor FastAPI
  │                              │
  │  POST /token                 │
  │  {username, password} ───────►  1. Validar credenciales
  │                              │  2. Generar JWT
  │  ◄─────────── {token}        │
  │                              │
  │  GET /usuarios/me            │
  │  Header: Bearer <token> ─────►  3. OAuth2PasswordBearer lee header
  │                              │  4. Validar token
  │  ◄─────────── {user_data}    │  5. Retornar datos
```

### Configuración de OAuth2PasswordBearer

In [42]:
# Esquema de seguridad OAuth2
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

# tokenUrl="token" indica que el endpoint de login es POST /token
print("✅ OAuth2PasswordBearer configurado")
print("  - El cliente debe enviar: Authorization: Bearer <token>")
print("  - El endpoint de login es: POST /token")

✅ OAuth2PasswordBearer configurado
  - El cliente debe enviar: Authorization: Bearer <token>
  - El endpoint de login es: POST /token


### Base de datos de usuarios (fake)

En producción esto sería una base de datos real. Por ahora usamos un diccionario.

In [43]:
# Base de datos fake con usuarios
# En producción: usar PostgreSQL, MongoDB, etc.
FAKE_USERS_DB = {
    "john": {
        "username": "john",
        "email": "john@example.com",
        "hashed_password": hash_password("secret123"),  # Password: secret123
        "disabled": False
    },
    "alice": {
        "username": "alice",
        "email": "alice@example.com",
        "hashed_password": hash_password("alicepass"),  # Password: alicepass
        "disabled": False
    }
}

print("✅ Base de datos de usuarios creada:")
for username in FAKE_USERS_DB.keys():
    print(f"  - {username}")

✅ Base de datos de usuarios creada:
  - john
  - alice


### Modelos Pydantic para autenticación

In [44]:
# Modelo para el token de respuesta
class Token(BaseModel):
    access_token: str
    token_type: str

# Modelo para datos del usuario
class User(BaseModel):
    username: str
    email: str
    disabled: bool = False

print("✅ Modelos Pydantic definidos")

✅ Modelos Pydantic definidos


### Endpoint POST /token (Login)

Este endpoint:
1. Recibe username y password en un formulario
2. Valida las credenciales
3. Genera y retorna un JWT

In [45]:
app = FastAPI()

@app.post("/token", response_model=Token)
async def login(form_data: OAuth2PasswordRequestForm = Depends()):
    """
    Endpoint de login que genera un token JWT.
    
    Recibe:
        - username (en formulario)
        - password (en formulario)
    
    Retorna:
        - access_token: JWT firmado
        - token_type: "bearer"
    """
    # 1. Buscar usuario en BD
    user = FAKE_USERS_DB.get(form_data.username)
    
    # 2. Validar que existe
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Usuario o contraseña incorrectos",
            headers={"WWW-Authenticate": "Bearer"},
        )
    
    # 3. Verificar contraseña
    if not verify_password(form_data.password, user["hashed_password"]):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Usuario o contraseña incorrectos",
            headers={"WWW-Authenticate": "Bearer"},
        )
    
    # 4. Generar token JWT
    access_token = crear_access_token(data={"sub": user["username"]})
    
    # 5. Retornar token
    return {"access_token": access_token, "token_type": "bearer"}

print("✅ Endpoint POST /token definido")

✅ Endpoint POST /token definido


### Prueba del endpoint /token

In [46]:
import uvicorn
import threading

client = TestClient(app)

# Login exitoso
response = client.post(
    "/token",
    data={"username": "john", "password": "secret123"}
)

assert response.status_code == 200
token_data = response.json()
print("✅ Login exitoso:")
print(f"  - Token: {token_data['access_token'][:50]}...")
print(f"  - Tipo: {token_data['token_type']}")

# Login con contraseña incorrecta
response_fail = client.post(
    "/token",
    data={"username": "john", "password": "incorrecta"}
)
assert response_fail.status_code == 401
print("\n❌ Login fallido correctamente rechazado")

# Servidor
print("\n")
run_api(app, port=8004)

✅ Login exitoso:
  - Token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJqb...
  - Tipo: bearer

❌ Login fallido correctamente rechazado


🚀 Servidor REINICIADO en http://127.0.0.1:8004
📄 Documentación: http://127.0.0.1:8004/docs


### Micro-reto: Formulario vs JSON

OAuth2PasswordRequestForm espera datos como **formulario** (Content-Type: application/x-www-form-urlencoded), NO como JSON.

In [55]:
# TODO: Intenta enviar las credenciales como JSON en vez de formulario
# ¿Qué error recibes?

response_json = client.post(
    "/token",
    json={"username": "john", "password": "secret123"}  # ❌ Esto NO funciona
)

print(f"Status: {response_json.status_code}")
print(f"Error: {response_json.json()}")
print("\n⚠️ OAuth2 requiere Content-Type: application/x-www-form-urlencoded")

Status: 422
Error: {'detail': [{'type': 'missing', 'loc': ['body', 'username'], 'msg': 'Field required', 'input': None}, {'type': 'missing', 'loc': ['body', 'password'], 'msg': 'Field required', 'input': None}]}

⚠️ OAuth2 requiere Content-Type: application/x-www-form-urlencoded


---

## 5. PROTECCIÓN DE RUTAS

### Dependencia: obtener_usuario_actual

Esta dependencia:
1. Recibe el token (vía OAuth2PasswordBearer)
2. Lo decodifica y valida
3. Busca el usuario en la BD
4. Retorna el usuario o lanza 401

In [56]:
async def obtener_usuario_actual(token: str = Depends(oauth2_scheme)) -> User:
    """
    Dependencia que valida el token y retorna el usuario actual.
    
    Args:
        token: Token JWT extraído del header Authorization
    
    Returns:
        Usuario autenticado
    
    Raises:
        HTTPException 401: Si el token es inválido o el usuario no existe
    """
    # 1. Decodificar token
    payload = decodificar_token(token)
    
    # 2. Extraer username del payload
    username: str = payload.get("sub")
    if username is None:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Token inválido",
            headers={"WWW-Authenticate": "Bearer"},
        )
    
    # 3. Buscar usuario en BD
    user_dict = FAKE_USERS_DB.get(username)
    if user_dict is None:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Usuario no encontrado",
            headers={"WWW-Authenticate": "Bearer"},
        )
    
    # 4. Retornar usuario
    return User(**user_dict)

# Tipo reutilizable
CurrentUserDep = Annotated[User, Depends(obtener_usuario_actual)]

print("✅ Dependencia obtener_usuario_actual definida")

✅ Dependencia obtener_usuario_actual definida


### Endpoint protegido: GET /usuarios/me

In [57]:
@app.get("/usuarios/me", response_model=User)
async def obtener_mi_perfil(current_user: CurrentUserDep):
    """
    Endpoint protegido que retorna el perfil del usuario autenticado.
    
    Requiere:
        - Header: Authorization: Bearer <token>
    
    Retorna:
        - Datos del usuario actual
    """
    return current_user

print("✅ Endpoint GET /usuarios/me definido (protegido)")

✅ Endpoint GET /usuarios/me definido (protegido)


### Prueba: Acceso con token válido

In [62]:
# 1. Login para obtener token
response_login = client.post(
    "/token",
    data={"username": "john", "password": "secret123"}
)
token = response_login.json()["access_token"]

# 2. Acceder a endpoint protegido con token
response_perfil = client.get(
    "/usuarios/me",
    headers={"Authorization": f"Bearer {token}"}
)

assert response_perfil.status_code == 200
user_data = response_perfil.json()
print("✅ Acceso autorizado:")
print(f"  - Usuario: {user_data['username']}")
print(f"  - Email: {user_data['email']}")

run_api(app)

✅ Acceso autorizado:
  - Usuario: john
  - Email: john@example.com
🚀 Servidor REINICIADO en http://127.0.0.1:8000
📄 Documentación: http://127.0.0.1:8000/docs


### Prueba: Acceso SIN token

In [59]:
# Intento sin header Authorization
response_sin_token = client.get("/usuarios/me")

assert response_sin_token.status_code == 401
print("❌ Acceso denegado (sin token):")
print(f"  - Status: {response_sin_token.status_code}")
print(f"  - Detail: {response_sin_token.json()['detail']}")

❌ Acceso denegado (sin token):
  - Status: 401
  - Detail: Not authenticated


### Prueba: Acceso con token inválido

In [60]:
# Intento con token falso
response_token_falso = client.get(
    "/usuarios/me",
    headers={"Authorization": "Bearer token-falso-123"}
)

assert response_token_falso.status_code == 401
print("❌ Acceso denegado (token inválido):")
print(f"  - Status: {response_token_falso.status_code}")
print(f"  - Detail: {response_token_falso.json()['detail']}")

app.openapi_schema = None
run_api(app)

❌ Acceso denegado (token inválido):
  - Status: 401
  - Detail: Token inválido o expirado
🚀 Servidor REINICIADO en http://127.0.0.1:8000
📄 Documentación: http://127.0.0.1:8000/docs


### Endpoints públicos vs protegidos

No todos los endpoints necesitan autenticación.

In [53]:
# Endpoint público (sin dependencia de autenticación)
@app.get("/")
async def root():
    return {"mensaje": "API pública"}

# Endpoint protegido (con CurrentUserDep)
@app.get("/privado")
async def privado(current_user: CurrentUserDep):
    return {"mensaje": "Datos privados", "usuario": current_user.username}

# Pruebas
print("✅ Acceso público (sin token):")
response_public = client.get("/")
print(f"  - Status: {response_public.status_code}")

print("\n❌ Acceso privado (sin token):")
response_private_no_token = client.get("/privado")
print(f"  - Status: {response_private_no_token.status_code}")

print("\n✅ Acceso privado (con token):")
response_private_with_token = client.get(
    "/privado",
    headers={"Authorization": f"Bearer {token}"}
)
print(f"  - Status: {response_private_with_token.status_code}")
print(f"  - Data: {response_private_with_token.json()}")

✅ Acceso público (sin token):
  - Status: 200

❌ Acceso privado (sin token):
  - Status: 401

✅ Acceso privado (con token):
  - Status: 200
  - Data: {'mensaje': 'Datos privados', 'usuario': 'john'}


---

### Endpoint de Registro: POST /register

**Completando el ciclo:** Hasta ahora solo podemos loguearnos con usuarios hardcodeados (john, alice).

Necesitamos un endpoint donde los usuarios puedan **registrarse** enviando su password en texto plano, y nosotros la guardamos hasheada.

In [54]:
# Modelo para registro
class UserRegister(BaseModel):
    username: str
    email: str
    password: str  # ⚠️ Viene en texto plano, la hashearemos

@app.post("/register", response_model=User)
async def registrar_usuario(user_data: UserRegister):
    """
    Endpoint de registro de nuevos usuarios.
    
    Recibe:
        - username: Nombre de usuario único
        - email: Email del usuario
        - password: Contraseña en texto plano
    
    Retorna:
        - Datos del usuario creado (sin la contraseña)
    """
    # 1. Validar que el usuario no existe
    if user_data.username in FAKE_USERS_DB:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="El usuario ya existe"
        )
    
    # 2. Hashear la contraseña
    hashed_password = hash_password(user_data.password)
    
    # 3. Guardar usuario en BD (fake)
    FAKE_USERS_DB[user_data.username] = {
        "username": user_data.username,
        "email": user_data.email,
        "hashed_password": hashed_password,
        "disabled": False
    }
    
    # 4. Retornar usuario creado (sin password)
    return User(
        username=user_data.username,
        email=user_data.email,
        disabled=False
    )

print("✅ Endpoint POST /register definido")

✅ Endpoint POST /register definido


### Prueba del endpoint /register

In [ ]:
# Registrar nuevo usuario
# Limpiamos la DB para que la prueba siempre funcione desde cero
FAKE_USERS_DB.clear() 
response_register = client.post(
    "/register",
    json={
        "username": "bob",
        "email": "bob@example.com",
        "password": "bobsecret123"
    }
)

assert response_register.status_code == 200
user_created = response_register.json()
print("✅ Usuario registrado:")
print(f"  - Username: {user_created['username']}")
print(f"  - Email: {user_created['email']}")

# Intentar registrar usuario duplicado
response_duplicate = client.post(
    "/register",
    json={
        "username": "bob",
        "email": "otro@example.com",
        "password": "otra"
    }
)
assert response_duplicate.status_code == 400
print("\n❌ Usuario duplicado rechazado correctamente")

# Probar login con usuario recién registrado
response_login_bob = client.post(
    "/token",
    data={"username": "bob", "password": "bobsecret123"}
)
assert response_login_bob.status_code == 200
print("\n✅ Login exitoso con usuario recién registrado")
print(f"  - Token generado: {response_login_bob.json()['access_token'][:30]}...")

app.openapi_schema = None
run_api(app, port=8005)

🚀 Servidor REINICIADO en http://127.0.0.1:8005
📄 Documentación: http://127.0.0.1:8005/docs


### Flujo completo: Register → Login → Acceso protegido

In [36]:
# FLUJO COMPLETO: Usuario nuevo desde cero

print("🔐 FLUJO COMPLETO DE AUTENTICACIÓN\n")

# 1. REGISTRO
print("1️⃣ REGISTRO")
response_reg = client.post(
    "/register",
    json={
        "username": "charlie",
        "email": "charlie@example.com",
        "password": "charlie123"
    }
)
print(f"   Status: {response_reg.status_code}")
print(f"   Usuario: {response_reg.json()['username']}\n")

# 2. LOGIN
print("2️⃣ LOGIN")
response_log = client.post(
    "/token",
    data={"username": "charlie", "password": "charlie123"}
)
token_charlie = response_log.json()["access_token"]
print(f"   Status: {response_log.status_code}")
print(f"   Token: {token_charlie[:40]}...\n")

# 3. ACCESO A RUTA PROTEGIDA
print("3️⃣ ACCESO A RUTA PROTEGIDA")
response_me = client.get(
    "/usuarios/me",
    headers={"Authorization": f"Bearer {token_charlie}"}
)
print(f"   Status: {response_me.status_code}")
print(f"   Usuario autenticado: {response_me.json()['username']}")
print(f"   Email: {response_me.json()['email']}")

print("\n✅ Flujo completo exitoso: Register → Login → Acceso protegido")

# run_api(app, port = 8003)

🔐 FLUJO COMPLETO DE AUTENTICACIÓN

1️⃣ REGISTRO
   Status: 200
   Usuario: charlie

2️⃣ LOGIN
   Status: 200
   Token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ...

3️⃣ ACCESO A RUTA PROTEGIDA
   Status: 200
   Usuario autenticado: charlie
   Email: charlie@example.com

✅ Flujo completo exitoso: Register → Login → Acceso protegido


### Resumen del flujo completo

```python
# 1. REGISTRO (hash de password)
@app.post("/register")
def register(user_data: UserRegister):
    hashed = hash_password(user_data.password)
    # Guardar en BD: user.hashed_password = hashed
    return user

# 2. LOGIN (verificar password y generar JWT)
@app.post("/token")
def login(form_data: OAuth2PasswordRequestForm):
    user = get_user(form_data.username)
    if not verify_password(form_data.password, user.hashed_password):
        raise HTTPException(401)
    token = crear_access_token({"sub": user.username})
    return {"access_token": token}

# 3. ACCESO A RUTAS PROTEGIDAS (validar JWT)
@app.get("/usuarios/me")
def perfil(current_user: CurrentUserDep):  # ← Dependencia valida token
    return current_user
```